sycl/doc/extensions/proposed/sycl_ext_oneapi_fpga_mem.asciidoc

```diff
@@ -0,0 +1,507 @@
+
+    T &get() noexcept;
+    const T &get() const noexcept;
+
+    // Enable assignments from underlying type
+    fpga_mem &operator=(const T &) noexcept;
+
+    // Note that there is no need for "fpga_mem" to define member functions
+    // for operators like "++", "[]", "->", comparison, etc. Instead, the type
+    // "T" need only define these operators as non-member functions. Because
+    // there is an implicit conversion from "fpga_mem" to "T&", the operations
+    // can be applied to objects of type "fpga_mem<T>"
+
+    template <typename propertyT> static constexpr bool has_property();
+
+    // The return type is an unspecified internal class used to represent
+    // instances of propertyT
+    template <typename propertyT>
+    static constexpr /*unspecified*/ get_property();
+  };
+
+} // namespace sycl::ext::intel::experimental
+----
+
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename ... S>
+constexpr fpga_mem_base(S ... args);
+----
+|
+Constructs a `fpga_mem` object, and implicit storage for `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template <access::decorated IsDecorated>
+multi_ptr<T, access::address_space::private_space, IsDecorated>
+  get_multi_ptr() noexcept;
+
+template <access::decorated IsDecorated>
+multi_ptr<T, access::address_space::private_space, IsDecorated>
+  get_multi_ptr() const noexcept;
+
+----
+|
+Returns a `multi_ptr` to the underlying `T` on the device.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator T&() noexcept;
+operator const T&() const noexcept;
+----
+|
+Implicit conversion to a reference to the underlying `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+T& get() noexcept;
+const T& get() const noexcept;
+----
+|
+Returns a reference to the underlying `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr bool has_property();
+----
+| Returns true if the `PropertyListT` contains the property specified by `propertyT`. Returns false if it does not.
+Available only if `sycl::is_property_key_of_v<propertyT, sycl::ext::intel::experimental::fpga_mem>` is true.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr auto get_property();
+----
+| Returns an object of the class used to represent the value of property `propertyT`.
+Must produce a compiler diagnostic if `PropertyListT` does not contain a `propertyT` property.
+Available only if `sycl::is_property_key_of_v<propertyT, sycl::ext::intel::experimental::fpga_mem>` is true.
+
+|===
+
+=== Restrictions on creating fpga_mem objects
+
+There are restrictions on how the application can create objects of type
+`fpga_mem`. Applications that violate these restrictions are ill-formed.
+
+* The `fpga_mem` variable must not itself be wrapped in a `fpga_mem` class
+* The `fpga_mem` variable must not have dynamic storage duration
```

Comment:
As above for fpga_datapath, we can probably get rid of these restrictions if we define fpga_mem right.